cmd/util/printer/job_progress_printer.go

```diff
@@ -0,0 +1,374 @@
+	jobProgressEventTopicCol,
+	jobProgressEventEventCol,
+}
+
+func NewJobProgressPrinter(client clientv2.API, runtimeSettings *cliflags.RunTimeSettings) *JobProgressPrinter {
+	return &JobProgressPrinter{
+		client:          client,
+		runtimeSettings: runtimeSettings,
+	}
+}
+
+// PrintJobProgress displays the job progress depending upon on cli runtime
+// settings
+func (j *JobProgressPrinter) PrintJobProgress(ctx context.Context, jobID string, cmd *cobra.Command) error {
+	// If we are in `--wait=false` print the jobID and then exit.
+	// All the code after this point is to show the progress of the job.
+	if !j.runtimeSettings.WaitForJobToFinish {
+		cmd.Print(jobID + "\n")
+		return nil
+	}
+
+	// If we are n `id-only` mode - print the id
+	if j.runtimeSettings.PrintJobIDOnly {
+		cmd.Print(jobID + "\n")
+	}
+
+	// Follow Logs
+	if j.runtimeSettings.Follow {
+		return j.followLogs(jobID, cmd)
+	}
+
+	// If we are only printing the id, set the rest of the output to "quiet",
+	// i.e. don't print
+	quiet := j.runtimeSettings.PrintJobIDOnly
+
+	jobErr := j.followProgress(ctx, jobID, cmd, quiet)
+	if jobErr != nil {
+		if jobErr.Error() == PrintoutCanceledButRunningNormally {
+			return nil
+		}
+
+		history, err := j.client.Jobs().History(ctx, &apimodels.ListJobHistoryRequest{
+			JobID:     jobID,
+			EventType: "execution",
+		})
+		if err != nil {
+			return fmt.Errorf("failed getting job history: %w", err)
+		}
+
+		historySummary := SummariseHistoryEvents(history.Items)
+		if len(historySummary) > 0 {
+			for _, event := range historySummary {
+				PrintEvent(cmd, event)
+			}
+		} else {
+			PrintError(cmd, jobErr)
+		}
+	}
+
+	if j.runtimeSettings.PrintNodeDetails || jobErr != nil {
+		executions, err := j.client.Jobs().Executions(ctx, &apimodels.ListJobExecutionsRequest{
+			JobID: jobID,
+		})
+		if err != nil {
+			return fmt.Errorf("failed getting job executions: %w", err)
+		}
+		summary := SummariseExecutions(executions.Items)
+		if len(summary) > 0 {
+			cmd.Println("\nJob Results By Node:")
+			for message, runs := range summary {
+				nodes := len(lo.Uniq(runs))
+				prefix := fmt.Sprintf("• Node %s: ", runs[0])
+				if len(runs) > 1 {
+					prefix = fmt.Sprintf("• %d runs on %d nodes: ", len(runs), nodes)
+				}
+				printIndentedString(cmd, prefix, strings.Trim(message, "\n"), none, 0)
+			}
+		} else {
+			cmd.Println()
+		}
+	}
+
+	if !quiet {
+		cmd.Println()
+		cmd.Println("To get more details about the run, execute:")
+		cmd.Println("\t" + os.Args[0] + " job describe " + jobID)
+
+		cmd.Println()
+		cmd.Println("To get more details about the run executions, execute:")
+		cmd.Println("\t" + os.Args[0] + " job executions " + jobID)
+	}
+
+	return nil
+}
+
+func (j *JobProgressPrinter) followLogs(jobID string, cmd *cobra.Command) error {
+	cmd.Printf("Job successfully submitted. Job ID: %s\n", jobID)
+	cmd.Printf("Waiting for logs... (Enter Ctrl+C to exit at any time, your job will continue running):\n\n")
```

Comment:
can you share snippets in the PR comment of different invocations of job submissions? For example, `id-only`, `wait=false`, follow logs, queueing triggered, queueing triggered and fulfilled, and cases with failures

Reply:
Acknowledged. Doing that now.

Reply:
Thanks. Looking at `--id-only` output, it should only be the id of the job so users can pipe it into other commands, such as `bacalhau docker run ubuntu echo hello --id-only | xargs bacalhau job describe`

It will wait until job completion, but only print the job id

Reply:
Makes sense ! Fixing it now :)

Reply:
Done with commit https://github.com/bacalhau-project/bacalhau/pull/4254/commits/57add2bc6343fbcb9fa299b96411f2ca8f0b6b46